### ISSUE TITLE
pandas.Index.to_numpy() method's arguments not properly typed

### ISSUE THREAD

**Issue by jbiz25** [pandas-dev/pandas-stubs#237]
**Describe the bug**
The pandas.Index.to_numpy() method takes arguments, but mypy complains about them, whether they are positional or keyword.

**To Reproduce**

import pandas as pd
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
array = df.index.to_numpy(dtype="int32")

This yields error: Unexpected keyword argument "dtype" for "to_numpy" of "Index"

import pandas as pd
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
array = df.index.to_numpy(dtype="int32")

This yields error: Too many arguments for "to_numpy" of "Index

**Please complete the following information:**
 - OS: Windows
 - OS Version 10
 - python version 3.10
 - version of type checker: latest
 - version of installed `pandas-stubs`1.4.3.220829

**Comment by Dr-Irv** [comment]
Thanks for the report.
Just need to fix:  https://github.com/pandas-dev/pandas-stubs/blob/80f561fc73f387f81f3876eeef4fe3bae82dae18/pandas-stubs/core/indexes/base.pyi#L207

PR and tests welcome!

**Comment by LeMeteore** [comment]
What about something like:

```python
from typing import Any, Optional
from pandas._typing import npt
import pandas._libs.lib as lib

    def to_numpy(                                                                                                          
            self, dtype: npt.DTypeLike | None = None,                                                                      
            copy: bool = False,                                                                                            
            na_value: Optional[Any] = lib.no_default,                                                                      
            **kwargs                                                                                                       
    ) -> np.ndarray: ...    
```
If good enough I can send the PR.

**Comment by jbiz25** [comment]
Thanks guys.  Also, FYI, it is probably worth investigating whether other the pandas objects that have to_numpy() methods are missing these, so that you can create similar annotations for them.

**Comment by Dr-Irv** [comment]
> If good enough I can send the PR.

For a stub, you want to have `= ...` rather than specify the default arguments.  

If you can make that change there, and as suggested above, look at the other `to_numpy()` calls in `pandas-stubs`, and also add tests, that would be a welcome PR!

**Comment by bashtage** [comment]
You can also run stubgen against pandas main to get the up to date list of
parameters and often correct type info.

**Comment by LeMeteore** [comment]
Very good catch. According to grep, there are many places where the fix should be applied:

```
>> grep -Hrnis "to_numpy" pandas-stubs/
pandas-stubs/_libs/tslibs/nattype.pyi:25:    def to_numpy(
pandas-stubs/_libs/tslibs/timedeltas.pyi:147:    def to_numpy(self) -> np.timedelta64: ...
pandas-stubs/_libs/tslibs/timestamps.pyi:203:    def to_numpy(
pandas-stubs/core/indexes/base.pyi:212:    def to_numpy(
pandas-stubs/core/series.pyi:1626:    def to_numpy(
pandas-stubs/core/arrays/masked.pyi:14:    def to_numpy(self, dtype=..., copy=..., na_value: Scalar = ...): ...
pandas-stubs/core/arrays/numpy_.pyi:112:    def to_numpy(self, dtype=..., copy: bool = ..., na_value=...): ...
pandas-stubs/core/arrays/base.pyi:15:    def to_numpy(self, dtype=..., copy: bool = ..., na_value=...): ...
pandas-stubs/core/frame.pyi:221:    def to_numpy(
pandas-stubs/core/base.pyi:44:    def to_numpy(self) -> np.ndarray: ...
```

If you guys agree that the type signature should be the same everywhere, I can work on it and send a PR in the next couple of hours.

**Comment by Dr-Irv** [comment]
> If you guys agree that the type signature should be the same everywhere, I can work on it and send a PR in the next couple of hours.

It definitely should not be the same everywhere, especially with respect to the type of the return values.  Many of the cases you indicate above are already typed.  You would need to look at the pandas sources, as well as the docs, to see where things are the same or different.

The argument types that are shared (`dtype` and `copy` and `na_value`) should be consistent.  Right now, I don't believe that is the case.  Also, you should take a look at each one and determine what the right values should be based on what is inside of pandas, and what the docs say.

For any signature that you update, you should also add a test.

Thanks for your desire to work on this!

**Comment by LeMeteore** [comment]
> The argument types that are shared (dtype and copy and na_value) should be consistent. Right now, I don't believe that is the case. 

Indeed, that is not the case:
- for the dtype parameter: sometimes `np.dtype`, sometimes `npt.DTypeLike`, sometimes `DtypeNp`,  `Dtype`. I went for `npt.DTypeLike`
- for the na_value parameter: sometimes `object` sometimes `Scalar`. I went for `Scalar`

> For any signature that you update, you should also add a test.

Do you mind please, giving more indications ? For [example](https://github.com/pandas-dev/pandas-stubs/blob/main/tests/test_frame.py#L779), I see 
```python
def test_types_to_numpy() -> None:                                                                                                                    
    df = pd.DataFrame(data={"col1": [1, 1, 2], "col2": [3, 4, 5]})                                                                                    
    df.to_numpy()                                                                                                                                     
    df.to_numpy(dtype="str", copy=True)                                                                                                               
    # na_value param was added in 1.1.0 https://pandas.pydata.org/docs/whatsnew/v1.1.0.html                                                           
    df.to_numpy(na_value=0) 
```
No call to `assert_type` ? 

Correct me if I'm wrong please, I should add similar functions: 
- for indexes, create an index, call to_numpy and assert the type
- for series, create a serie, call to_numpy and assert the type

Also. It seems like there is an addition [here](https://github.com/pandas-dev/pandas/blob/main/pandas/core/arrays/string_arrow.py#L163). Should I include the addition in the current PR or create a different one ?

Thanks for your time, advices and suggestions o/

**Comment by Dr-Irv** [comment]
With respect to the tests not having `check(assert_type(...))`, that's a legacy issue.  The original stubs didn't have them, but when we are adding new tests, we are requiring it.  Eventually, all the tests should use that paradigm.  Would be good if you fix all the `to_numpy()` tests to use that pattern.

> Correct me if I'm wrong please, I should add similar functions:
> 
> - for indexes, create an index, call to_numpy and assert the type
> - for series, create a serie, call to_numpy and assert the type

Yes, you should add similar functions for all the `to_numpy()` variants.  
> 
> Also. It seems like there is an addition [here](https://github.com/pandas-dev/pandas/blob/main/pandas/core/arrays/string_arrow.py#L163). Should I include the addition in the current PR or create a different one ?

You can do one PR, that includes fixes to all the `to_numpy()` stubs, and the appropriate tests.

Thanks in advance for your contribution!

**Comment by Dr-Irv** [comment]
closed via #270
